src/autoencoder.ts

```diff
@@ -0,0 +1,189 @@
+import { KernelOutput, Texture, TextureArrayOutput } from "gpu.js";
+import { IJSONLayer, INeuralNetworkData, INeuralNetworkDatum, INeuralNetworkTrainOptions } from "./neural-network";
+import { INeuralNetworkGPUOptions, NeuralNetworkGPU } from "./neural-network-gpu";
+import { INeuralNetworkState } from "./neural-network-types";
+import { UntrainedNeuralNetworkError } from "./errors/untrained-neural-network-error";
+
+export interface IAEOptions {
+  binaryThresh: number;
+  decodedSize: number;
+  hiddenLayers: number[];
+}
+
+/**
+ * An autoencoder learns to compress input data down to relevant features and reconstruct input data from its compressed representation.
+ */
+export class AE<DecodedData extends INeuralNetworkData, EncodedData extends INeuralNetworkData> {
+  private decoder?: NeuralNetworkGPU<EncodedData, DecodedData>;
```

Comment:
Can we make AE support non-gpu Neural network? Or is there a specific reason why we are tying this to GPU NNs only?